File: python/News-Article-Scraper/Python/news_article_scraper.py
from bs4 import BeautifulSoup
import requests
import csv


def parse_sitemap() -> list:
    response = requests.get("https://www.example.com/sitemap.xml")
    if response.status_code != 200:
        return None
    xml_as_str = response.text

    soup = BeautifulSoup(xml_as_str, "lxml")
    loc_elements = soup.find_all("loc")
    links = []
    for loc in loc_elements:
        links.append(loc.text)

    print(f'Found {len(links)} links')
    return links


def parse_articles(links: list):
    s = requests.Session()
    with open("news.csv", "w", encoding="utf-8", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=['Heading', 'Body'])
        writer.writeheader()
        for link in links:
            response = s.get(link)
            soup = BeautifulSoup(response.text, "lxml")
            heading = soup.select_one('h1').text
            para = []
            for p in soup.select('.complete-story p'):
                para.append(p.text)
            body = '\n'.join(para)
            writer.writerow({'Heading': heading,
                             'Body': body
                             })


if __name__ == '__main__':
    links = parse_sitemap()
    parse_articles(links)

File: python/News-Article-Scraper/JavaScript/news_article_scraper.js
const cheerio = require("cheerio");
const axios = require("axios");
url = `https://www.example.com/sitemap.xml`;
let links = [];
async function getLinks() {
    try {
        const response = await axios.get(url);
        const $ = cheerio.load(response.data, { xmlMode: true });
        all_loc = $('loc');
        all_loc.each(function () {
            links.push($(this).text());
        })
        console.log(links.length + ' links found.');
        links.forEach(async function (story_link) {
            try {
                let story = await axios.get(story_link);
                let $ = cheerio.load(story.data);
                heading = $('h1').text()
                body = $('.complete-story p').text()

            } catch (error) {
                console.error('internal\n' + error)
            }
        })

    } catch (error) {
        console.error(error);
    }
}
getLinks();
